src/Workspaces/Core/MSBuild.BuildHost/BuildHost.cs

```diff
@@ -43,36 +43,61 @@ private bool TryEnsureMSBuildLoaded(string projectOrSolutionFilePath)
-            instance = MSBuildLocator.QueryVisualStudioInstances().OrderByDescending(vs => vs.Version).FirstOrDefault();
+                // TODO: we should also check that the managed tools are actually installed
+                instance = MSBuildLocator.QueryVisualStudioInstances().OrderByDescending(vs => vs.Version).FirstOrDefault();
 
 #else
 
-            // Locate the right SDK for this particular project; MSBuildLocator ensures in this case the first one is the preferred one.
-            // TODO: we should pick the appropriate instance back in the main process and just use the one chosen here.
-            var options = new VisualStudioInstanceQueryOptions { DiscoveryTypes = DiscoveryType.DotNetSdk, WorkingDirectory = Path.GetDirectoryName(projectOrSolutionFilePath) };
-            instance = MSBuildLocator.QueryVisualStudioInstances(options).FirstOrDefault();
+                // Locate the right SDK for this particular project; MSBuildLocator ensures in this case the first one is the preferred one.
+                // TODO: we should pick the appropriate instance back in the main process and just use the one chosen here.
+                var options = new VisualStudioInstanceQueryOptions { DiscoveryTypes = DiscoveryType.DotNetSdk, WorkingDirectory = Path.GetDirectoryName(projectOrSolutionFilePath) };
+                instance = MSBuildLocator.QueryVisualStudioInstances(options).FirstOrDefault();
 
 #endif
 
-            if (instance != null)
-            {
-                MSBuildLocator.RegisterInstance(instance);
-                _logger.LogInformation($"Registered MSBuild instance at {instance.MSBuildPath}");
-                return true;
+                if (instance != null)
+                {
+                    MSBuildLocator.RegisterInstance(instance);
+                    _logger.LogInformation($"Registered MSBuild instance at {instance.MSBuildPath}");
+                }
+                else
+                {
+                    _logger.LogCritical("No compatible MSBuild instance could be found.");
+                }
             }
             else
             {
-                _logger.LogCritical("No compatible MSBuild instance could be found.");
-                return false;
+#if NETFRAMEWORK
+
+                // We're running on Mono, but not all Mono installations have a usable MSBuild installation, so let's see if we have one that we can use.
+                var monoMSBuildDirectory = MonoMSBuildDiscovery.GetMonoMSBuildDirectory();
+
+                if (monoMSBuildDirectory != null)
+                {
+                    MSBuildLocator.RegisterMSBuildPath(monoMSBuildDirectory);
+                    _logger.LogInformation($"Registered MSBuild instance at {monoMSBuildDirectory}");
+                }
+                else
+                {
+                    _logger.LogCritical("No Mono MSBuild installation could be found; see https://www.mono-project.com/ for installation instructions.");
```

Comment:
This url doesn't seem to be particularly helpful.  Wonder if we should have our own docs here describing how to acquire either mono+msbuild or .netframework msbuild tools (similar maybe to what is described in https://github.com/dotnet/vscode-csharp/issues/5120 )

Maybe not at this line exactly, but as mentioned before some unified pop-up when we cant find the .netframework / mono install would be good to have.

Reply:
The Omnisharp instructions weren't much better:

https://github.com/OmniSharp/omnisharp-roslyn/blob/dde8119c40f4e3920eb5ea894cbca047033bd9aa/src/OmniSharp.Host/MSBuild/Discovery/Providers/MonoInstanceProvider.cs#L95

I felt bad linking directly to a specific page so just restricted it to the domain. Do we want to own writing some docs here?

> but as mentioned before some unified pop-up when we cant find the .netframework / mono install would be good to have.

My hope was if we do a fallback, and it still fails, the user will still know because of the error from the fallback build. (i.e. you'll get build tasks failing because it's the wrong .NET, etc.) So at that point then users would still see it. I'd be worried about too many false positives where we detect a project as .NET Framework (even if it's not really) and then raise an error about a fallback when indeed the .NET Core option would have worked just fine.

Reply:
> I felt bad linking directly to a specific page so just restricted it to the domain. Do we want to own writing some docs here?

I think we absolutely should write our own docs here for the C# extension- lets file an issue for it (cc @mikadumont )

> My hope was if we do a fallback, and it still fails, the user will still know because of the error from the fallback build. (i.e. you'll get build tasks failing because it's the wrong .NET, etc.)

They'll get an error - but I don't know if they'll have enough information from that to be able to fix it (I guess if they look at the logs they'll see a warning).  Maybe if we detect it is .net framework AND we can't find the install AND it fails to load on .net core, we modify the error pop-up to ask them to install the .net framework or mono tools?

Reply:
> Maybe if we detect it is .net framework AND we can't find the install AND it fails to load on .net core, we modify the error pop-up to ask them to install the .net framework or mono tools?

That seems reasonable!

Reply:
@dibarbet Feelings about me replacing that with an aka.ms link that points to a generic download page that we'll later switch to the docs, or introduce the link only once needed?

Reply:
> @dibarbet Feelings about me replacing that with an aka.ms link that points to a generic download page that we'll later switch to the docs, or introduce the link only once needed?

The former sounds nice, but not bothered if we wait to do it when we actually write the docs.

Reply:
I've updated the logging to display a distinct message if we had an error when we also failed to find Mono/.NET Framework MSBuild.

Reply:
And bug filed to track better docs: https://github.com/dotnet/roslyn/issues/70281